### ISSUE TITLE
No input sample provided

### ISSUE THREAD

**Issue by JayFoxRox** [XboxDev/nxdk#177]
nxdk currently doesn't have an input sample.

There was a proposal in #106 but it was closed due to quality issues.

A new sample should be designed.
Ideally, this sample uses the stable SDL APIs (see below); it should showcase:

- Gamepad input state on the different controller ports
- Hotplugging events (currently unstable in USB driver / underlying API)
- Gamepad vibration (currently missing in USB driver / underlying API / SDL Haptic)
- XID type detection (currently missing in USB driver / underlying API)

Due to the complexity of this sample (4 ports + mix of state and event log), it's probably a good idea to make a small renderer using SDL (also to avoid use of the legacy debug API #172).

---

For anyone who wants to use input, we recommend using the [SDL_GameController API](https://wiki.libsdl.org/CategoryGameController) which is implemented in nxdk's SDL port.

Additionally, the SDL event API can be used to receive hotplug events such as [SDL_ControllerDeviceEvent](https://wiki.libsdl.org/SDL_ControllerDeviceEvent?highlight=%28%5CbCategoryStruct%5Cb%29%7C%28CategoryEvents%29).

You can also use the underlying [SDL_Joystick API](https://wiki.libsdl.org/CategoryJoystick), but the current SDL backend is a limiting factor and doesn't offer benefits over the SDL_GameController API.

Using the platform specific OpenXDK-XInput API exposed by  `<hal/input.h>` and `<hal/pad.h>` is not recommended. It is possible that this code will be replaced by a different API in the future.

**Comment by gbrown5** [comment]
I can't seem to get input working at all in SDL. 
`	
int frame = 0;    
while (!done) {

                //quick way to move text slowly up the screen 
		frame++;
		if(frame > 25)
		{
			frame = 0;
			fontY--;
			
			if(fontY < 20) fontY = SCREEN_HEIGHT - 100; //reset text's position
		}
		
        /* Check for events */
        while (SDL_PollEvent(&event)) {
            switch (event.type) {
            case SDL_QUIT:
                done = 1;
                break;

		case SDL_CONTROLLERAXISMOTION: 
			fontY-=25;
			break;
		
		case SDL_CONTROLLERBUTTONDOWN:
			fontY-= 25;
			break;
		
		case SDL_CONTROLLERBUTTONUP: 
			fontY-= 25;
			break;`
Note that I am rendering text later on using the variable 'fontY' for the Y position. When ran, the text slowly moves up at regular intervals 1px at a time but nothing happens when pressing buttons/moving joy sticks.

**Comment by dracc** [comment]
@gbrown5 Good thing about sdl2 is that it's cross platform for most parts.
Without more knowledge about your code base, the snippet you've shown looks *very* broken to me.

**Comment by gbrown5** [comment]
It's essentially a hello world example. I also compiled and ran the 'example' script located in "nxdk/lib/sdl/SDL2/test/testgamecontroller.c" and that doesn't work either...

**Comment by gbrown5** [comment]
Edit: Thank you JayFox, I guess the poll event was the problem. It didn't seem to work with the background input hint either. Moving to state based works.

**Comment by dracc** [comment]
I'd argue this was resolved by 176de1b .

**Comment by JayFoxRox** [comment]
Yes, partially. Ideally we'd also showcase the rest (multiple gamepads, hotplug, vibration, XID device type).
I believe the backend / USB code was changed in the past and *should* support all of those features by now?
